### PR TITLE
Handle typosquat restrictions

### PR DESCRIPTION
**Pull request by bconnorwhite**
Fixes https://github.com/sindresorhus/npm-name/issues/15

This solution should account for punctuation using [is-name-taken](https://github.com/bconnorwhite/is-name-taken), which caches the bulk of npm package names so that only the last few need to be fetched.

One thing to note is that npmName.many is kinda slow as it causes multiple requests to sync the list of packages. Instead, the list of names should really all be sent off to is-name-taken together. If anyone wants, I can add this as well.

### CODE REVIEW

**Comment by bconnorwhite** [comment]
Also, this doesn't account for punctuation in org names on their own (@orgname/example-pkg would work, but not @orgname). I can add a separate package to cache a list of org names if needed.

(I'm just assuming this is how org names work, the blog post doesn't really mention punctuation with respect to org names: https://blog.npmjs.org/post/168978377570/new-package-moniker-rules)

**Comment by sindresorhus** [comment]
> which caches the bulk of npm package names so that only the last few need to be fetched.

How is it cached and how is it kept fresh?

File: index.js [hunk 1]
```diff
@@ -43,17 +41,27 @@ const request = async (name, options) => {
 	}
 
 	try {
+		let urlName = name;
+		if (isScopedPackage) {
+			urlName = name.replace(/\//g, '%2f');
+		}
```
Comment (sindresorhus) [inline index.js:47]: Why is this moved inside the try/catch?

**Comment by sindresorhus** [comment]
> One thing to note is that npmName.many is kinda slow as it causes multiple requests to sync the list of packages.

Sounds like that should be fixed in `is-name-taken`.

**Comment by sindresorhus** [comment]
Bump

File: index.js (hunk 1 above)
Comment (bconnorwhite) [reply index.js]: Moved this back to keep consistent with original code: https://github.com/sindresorhus/npm-name/pull/42/commits/447f06a04fff8c68a56c2ef4060e2a6a69a0c428

**Comment by bconnorwhite** [comment]
When `isTaken` is called, I sync the list of packages first. Since I have a pointer to the last package seen, its faster to sync the more recent the list of package names is.

I also set up an action to publish this list daily, so the list will be at most 24 hrs behind on the latest version.

Alternatively, I added an "optimistic" flag (Ex: `isTaken("pkg-name", { optimistic: true })`), which skips the sync, if the risk of a package with a conflicting name being published since the last sync is tolerable.

**Comment by bconnorwhite** [comment]
Added a `maxAge` flag for more granular control, I'm seeing the tests run ~3x faster after setting this to 60000ms.

**Comment by sindresorhus** [comment]
Thanks :)
